### PR TITLE
SLIM-710: Deletes a scheduled task on failure

### PR DESCRIPTION
**Pull request by robindenadel**

### CODE REVIEW

File: osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java [hunk 1]
```diff
@@ -116,8 +116,8 @@ private boolean shouldRetryBasedOnException(final OsgpException e) {
     }
 
     private void handleScheduledTask(final ProtocolResponseMessage message) {
```
Comment (robindenadel) [inline osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java]: agree

File: osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java [hunk 2]
```diff
@@ -127,26 +127,20 @@ private void handleScheduledTask(final ProtocolResponseMessage message) {
...
-            this.scheduledTaskRepository.delete(scheduledTask);
             this.domainResponseMessageSender.send(message);
         } else {
-            final String errorMessage = message.getOsgpException() == null ? ""
-                    : message.getOsgpException().getCause() == null ? message.getOsgpException().getMessage() : message
-                            .getOsgpException().getCause().getMessage();
-            scheduledTask.setFailed(errorMessage);
 
             if (message.getRetryHeader() != null && message.getRetryHeader().shouldRetry()) {
                 scheduledTask.retryOn(message.getRetryHeader().getScheduledRetryTime());
```
Comment (robindenadel) [inline osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java]: you're right, will change that

File: osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java [hunk 3]
```diff
@@ -127,26 +127,20 @@ private void handleScheduledTask(final ProtocolResponseMessage message) {
...
             } else {
                 this.domainResponseMessageSender.send(message);
             }
-
-            this.scheduledTaskRepository.save(scheduledTask);
```
Comment (robindenadel) [inline osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java]: you're right, will change that

File: osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java [hunk 4]
```diff
@@ -116,38 +116,65 @@ private boolean shouldRetryBasedOnException(final OsgpException e) {
...
+
+    private boolean messageIsSuccessful(final ProtocolResponseMessage message, final ScheduledTask scheduledTask) {
+        return message.getResult() == ResponseMessageResultType.OK
+                && scheduledTask.getStatus() == ScheduledTaskStatusType.PENDING;
+    }
+
+    private String determineErrorMessage(final ProtocolResponseMessage message) {
+        final String errorMessage;
+        if (message.getOsgpException() == null) {
+            errorMessage = "";
```
Comment (hannek24) [inline osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java]: directly return "" is better

File: osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java [hunk 5]
```diff
@@ -116,38 +116,65 @@ private boolean shouldRetryBasedOnException(final OsgpException e) {
...
+        } else if (message.getOsgpException().getCause() == null) {
+            errorMessage = message.getOsgpException().getMessage();
```
Comment (hannek24) [inline osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java]: return directly

File: osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java [hunk 6]
```diff
@@ -116,38 +116,65 @@ private boolean shouldRetryBasedOnException(final OsgpException e) {
...
+            this.handleMessageRetry(message, scheduledTask);
+        } else {
+            this.domainResponseMessageSender.send(message);
+            this.scheduledTaskRepository.delete(scheduledTask);
+        }
+    }
+
+    private void handleMessageRetry(final ProtocolResponseMessage message, final ScheduledTask scheduledTask) {
+        final String errorMessage = this.determineErrorMessage(message);
+        scheduledTask.setFailed(errorMessage);
```
Comment (hannek24) [inline osgp-core/src/main/java/com/alliander/osgp/core/application/services/DeviceResponseMessageService.java]: Replace these two lines with:
`scheduledTask.setFailed(this.determineErrorMessage(message));`

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+    /**
+     * test processMessage with a scheduled task that has been successfull
+     */
+    @Test
+    public void testProcessScheduledMessageSuccess() {
+        final ResponseMessageResultType result = ResponseMessageResultType.OK;
+
+        final ProtocolResponseMessage message = new ProtocolResponseMessage.Builder()
+                .deviceMessageMetadata(DEVICE_MESSAGE_DATA).domain(DOMAIN).domainVersion(DOMAIN_VERSION).result(result)
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: directly add ResponseMessageResultType.OK in steadof 'result'

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    public void initTest() {
+        this.getMaxRetryCount = 1;
+    }
+
+    /**
+     * test processMessage with a scheduled task that has been successfull
+     */
+    @Test
+    public void testProcessScheduledMessageSuccess() {
+        final ResponseMessageResultType result = ResponseMessageResultType.OK;
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: can be removed

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+        final ProtocolResponseMessage message = new ProtocolResponseMessage.Builder()
+                .deviceMessageMetadata(DEVICE_MESSAGE_DATA).domain(DOMAIN).domainVersion(DOMAIN_VERSION).result(result)
+                .dataObject(DATA_OBJECT).scheduled(true).build();
+        final ScheduledTask scheduledTask = new ScheduledTask(DEVICE_MESSAGE_DATA, DOMAIN, DOMAIN, DATA_OBJECT,
+                SCHEDULED_TIME);
+        scheduledTask.setPending();
+        when(this.scheduledTaskRepository.findByCorrelationUid(anyString())).thenReturn(scheduledTask);
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: for findByCorrelationUid you can be as specific as possible. Not just any string, but the exact string "correlationId" is meant here.

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        // should be retried
+        final RetryHeader retryHeader = new RetryHeader(0, this.getMaxRetryCount, scheduledRetryTime);
+
+        final ProtocolResponseMessage message = new ProtocolResponseMessage.Builder()
+                .deviceMessageMetadata(DEVICE_MESSAGE_DATA).domain(DOMAIN).domainVersion(DOMAIN_VERSION).result(result)
+                .dataObject(DATA_OBJECT).scheduled(true).retryHeader(retryHeader).build();
+        final ScheduledTask scheduledTask = new ScheduledTask(DEVICE_MESSAGE_DATA, DOMAIN, DOMAIN, DATA_OBJECT,
+                SCHEDULED_TIME);
+
+        when(this.scheduledTaskRepository.findByCorrelationUid(anyString())).thenReturn(scheduledTask);
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: for findByCorrelationUid you can be as specific as possible. Not just any string, but the exact string "correlationId" is meant here.

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        final RetryHeader retryHeader = new RetryHeader(this.getMaxRetryCount, this.getMaxRetryCount,
+                scheduledRetryTime);
+
+        final ProtocolResponseMessage message = new ProtocolResponseMessage.Builder()
+                .deviceMessageMetadata(DEVICE_MESSAGE_DATA).domain(DOMAIN).domainVersion(DOMAIN_VERSION).result(result)
+                .dataObject(DATA_OBJECT).scheduled(true).retryHeader(retryHeader).build();
+        final ScheduledTask scheduledTask = new ScheduledTask(DEVICE_MESSAGE_DATA, DOMAIN, DOMAIN, DATA_OBJECT,
+                SCHEDULED_TIME);
+
+        when(this.scheduledTaskRepository.findByCorrelationUid(anyString())).thenReturn(scheduledTask);
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: for findByCorrelationUid you can be as specific as possible. Not just any string, but the exact string "correlationId" is meant here.

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        this.deviceResponseMessageService.processMessage(message);
+
+        // check if message is send and task is deleted
+        verify(this.domainResponseMessageSender, times(1)).send(message);
+        verify(this.scheduledTaskRepository, times(1)).delete(scheduledTask);
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: For both lines: times(1) isn't really necessary. It's the default.

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+                .dataObject(DATA_OBJECT).scheduled(true).retryHeader(retryHeader).build();
+        final ScheduledTask scheduledTask = new ScheduledTask(DEVICE_MESSAGE_DATA, DOMAIN, DOMAIN, DATA_OBJECT,
+                SCHEDULED_TIME);
+
+        when(this.scheduledTaskRepository.findByCorrelationUid(anyString())).thenReturn(scheduledTask);
+        this.deviceResponseMessageService.processMessage(message);
+
+        // check if the message is not send and the task is not deleted
+        verify(this.domainResponseMessageSender, never()).send(message);
+        verify(this.scheduledTaskRepository, never()).delete(scheduledTask);
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: using verifyNoMoreInteractions in shorter. like:
`verifyNoMoreInteractions(this.domainResponseMessageSender, this.scheduledTaskRepository);`

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        final ProtocolResponseMessage message = new ProtocolResponseMessage.Builder()
+                .deviceMessageMetadata(DEVICE_MESSAGE_DATA).domain(DOMAIN).domainVersion(DOMAIN_VERSION).result(result)
+                .dataObject(DATA_OBJECT).scheduled(true).retryHeader(retryHeader).build();
+        final ScheduledTask scheduledTask = new ScheduledTask(DEVICE_MESSAGE_DATA, DOMAIN, DOMAIN, DATA_OBJECT,
+                SCHEDULED_TIME);
+
+        when(this.scheduledTaskRepository.findByCorrelationUid(anyString())).thenReturn(scheduledTask);
+        this.deviceResponseMessageService.processMessage(message);
+
+        // check if the message is not send and the task is not deleted
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: You should also check if the scheduled task is saved. like:
`verify(this.scheduledTaskRepository).save(scheduledTask);`

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        calendar.add(Calendar.DATE, 1);
+        final Date scheduledRetryTime = calendar.getTime();
+
+        // since the retryCount is smaller than the maxRetries, the message
+        // should be retried
+        final RetryHeader retryHeader = new RetryHeader(0, this.getMaxRetryCount, scheduledRetryTime);
+
+        final ProtocolResponseMessage message = new ProtocolResponseMessage.Builder()
+                .deviceMessageMetadata(DEVICE_MESSAGE_DATA).domain(DOMAIN).domainVersion(DOMAIN_VERSION).result(result)
+                .dataObject(DATA_OBJECT).scheduled(true).retryHeader(retryHeader).build();
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: also set the osgpException for the retry and check it afterwards if it is indeed part of the scheduled message

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        calendar.add(Calendar.DATE, 1);
+        final Date scheduledRetryTime = calendar.getTime();
+
+        // since the retryCount equals the maxRetries, the message should fail
+        final RetryHeader retryHeader = new RetryHeader(this.getMaxRetryCount, this.getMaxRetryCount,
+                scheduledRetryTime);
+
+        final ProtocolResponseMessage message = new ProtocolResponseMessage.Builder()
+                .deviceMessageMetadata(DEVICE_MESSAGE_DATA).domain(DOMAIN).domainVersion(DOMAIN_VERSION).result(result)
+                .dataObject(DATA_OBJECT).scheduled(true).retryHeader(retryHeader).build();
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: also add the osgpException and check it afterwards if it is part of the message send back to the client

File: osgp-core/src/main/java/com/alliander/osgp/core/application/tasks/ScheduledTaskScheduler.java [hunk 17]
```diff
@@ -50,8 +50,8 @@ public void run() {
     private void processScheduledTasks(final ScheduledTaskStatusType type) {
```
Comment (hannek24) [inline osgp-core/src/main/java/com/alliander/osgp/core/application/tasks/ScheduledTaskScheduler.java]: Also a unit test needed to test this private method. Test needed like: when an FunctionalException occurs check that the task is deleted and a message is send back to the client

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+    @Mock
+    private DeviceRequestMessageService deviceRequestMessageService;
+
+    @Mock
+    private DeviceRepository deviceRepository;
+
+    @InjectMocks
+    private DeviceResponseMessageService deviceResponseMessageService;
+
+    @Autowired
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: Autowired is not needed

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+    private DomainResponseService domainResponseMessageSender;
+
+    @Mock
+    private ScheduledTaskRepository scheduledTaskRepository;
+
+    @Mock
+    private DeviceRequestMessageService deviceRequestMessageService;
+
+    @Mock
+    private DeviceRepository deviceRepository;
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: This mock can be left out

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+public class DeviceResponseMessageServiceTest {
+
+    @Mock
+    private DomainResponseService domainResponseMessageSender;
+
+    @Mock
+    private ScheduledTaskRepository scheduledTaskRepository;
+
+    @Mock
+    private DeviceRequestMessageService deviceRequestMessageService;
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: This mock is not needed.

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+
+    @Mock
+    private DeviceRepository deviceRepository;
+
+    @InjectMocks
+    private DeviceResponseMessageService deviceResponseMessageService;
+
+    @Autowired
+    private int getMaxRetryCount;
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: Directly assign with 1 and you can make a constant of it or leave it out completely and use the (non magic number) 1 everywhere

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+            "organisationId", "correlationId", "messageType", 4);
+    private static final String DOMAIN = "Domain";
+    private static final String DOMAIN_VERSION = "1.0";
+    private static final String DATA_OBJECT = "data object";
+    private static final Timestamp SCHEDULED_TIME = new Timestamp(Calendar.getInstance().getTime().getTime());
+
+    @Before
+    public void initTest() {
+        this.getMaxRetryCount = 1;
+    }
```
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: Entire before method is not needed

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java (hunk 8 above)
Comment (robindenadel) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: is needed to test the messageIsSuccessful branch

File: osgp-core/src/main/java/com/alliander/osgp/core/application/tasks/ScheduledTaskScheduler.java [hunk 23]
```diff
@@ -63,8 +63,7 @@ private void processScheduledTasks(final ScheduledTaskStatusType type) {
                 this.deviceRequestMessageService.processMessage(protocolRequestMessage);
             } catch (final FunctionalException e) {
                 LOGGER.error("Processing scheduled task failed.", e);
-                scheduledTask.setFailed(e.getMessage());
-                scheduledTask = this.scheduledTaskRepository.save(scheduledTask);
+                this.scheduledTaskRepository.delete(scheduledTask);
```
Comment (hannek24) [inline osgp-core/src/main/java/com/alliander/osgp/core/application/tasks/ScheduledTaskScheduler.java]: Yes, you're right

File: osgp-core/src/main/java/com/alliander/osgp/core/application/tasks/ScheduledTaskScheduler.java (hunk 17 above)
Comment (hannek24) [inline osgp-core/src/main/java/com/alliander/osgp/core/application/tasks/ScheduledTaskScheduler.java]: Yes, you're right

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java (hunk 8 above)
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: What I meant is that this value can be used directly on line 83. See also my comment for line 83

Comment (robindenadel) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: ok, clear now

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java (hunk 10 above)
Comment (robindenadel) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: correlationId is not important in this test, no need to be as specific as you can

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java (hunk 16 above)
Comment (robindenadel) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: for this branch, nothing is done with the osgpException

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java (hunk 11 above)
Comment (robindenadel) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: correlationId is not important in this test, no need to be as specific as you can

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java (hunk 13 above)
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: ok

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java (hunk 11 above)
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: ok

File: osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java (hunk 16 above)
Comment (hannek24) [inline osgp-core/src/test/java/com/alliander/osgp/core/application/services/DeviceResponseMessageServiceTest.java]: agree

**Pull request merged**